File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.py
import heapq
def solution(jobs):
    start = -1
    answer, fin_job, curr = 0, 0, 0
    waiting = []
    jobs.sort()
    while fin_job < len(jobs):
        for job in jobs:
            if start < job[0] <= curr:
                heapq.heappush(waiting, (job[1], job[0])) #(소요시간, 시작시간)
        if len(waiting) == 0:
            curr += 1
            continue
        job = heapq.heappop(waiting)
        start = curr
        curr += job[0]
        answer += (curr - job[1])
        fin_job += 1
        
    return answer // len(jobs)

File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.java
import java.util.*;
class Solution {
    public int solution(int[][] jobs) {
        int answer = 0;  
        Arrays.sort(jobs, (o1, o2) -> Integer.compare(o1[0], o2[0]));
        PriorityQueue<int[]> pq = new PriorityQueue<>((o1, o2) -> Integer.compare(o1[0], o2[0]));
        int curr_time = 0;
        int curr_idx = 0;
        int cnt = 0;
        while (cnt < jobs.length) {
            while (curr_idx < jobs.length && jobs[curr_idx][0] <= curr_time) {
                pq.add(new int[]{jobs[curr_idx][1], jobs[curr_idx][0]}); //(걸리는 시간, 시작 시간)
                curr_idx += 1;
            }
            if (!pq.isEmpty()) {
                int[] time = pq.poll();
                cnt += 1;
                answer += (curr_time - time[1] + time[0]);
                curr_time += time[0];
            } else {
                curr_time += 1;
            }
        }
        
        return answer / jobs.length;
    }
}